python_modules/libraries/dagster-managed-elements/dagster_managed_elements/types.py

```diff
@@ -0,0 +1,209 @@
+import enum
+from abc import ABC, abstractmethod
+from typing import Any, List, NamedTuple, Optional, OrderedDict, Tuple, Union
+
+import click
+
+
+class ManagedElementError(enum.Enum):
+    CANNOT_CONNECT = "cannot_connect"
+
+
+SANITIZE_KEY_KEYWORDS = ["password", "token", "secret"]
+
+
+def _sanitize(key: str, value: str):
+    """
+    Rudamentary sanitization of values so we can avoid printing passwords
+    to the console.
+    """
+    if any(keyword in key for keyword in SANITIZE_KEY_KEYWORDS):
+        return "**********"
+    return value
+
+
+class ManagedElementDiff(
+    NamedTuple(
+        "_ManagedElementDiff",
+        [
+            ("additions", List[Tuple[str, str]]),
+            ("deletions", List[Tuple[str, str]]),
+            ("modifications", List[Tuple[str, str, str]]),
+            ("nested", OrderedDict[str, "ManagedElementDiff"]),
+        ],
+    )
+):
+    """
+    Utility class representing the diff between configured and deployed managed element. Can be rendered to a
+    color-coded, user-friendly string.
+    """
+
+    def __new__(
+        cls,
+        additions: Optional[List[Tuple[str, str]]] = None,
+        deletions: Optional[List[Tuple[str, str]]] = None,
+        modifications: Optional[List[Tuple[str, str, str]]] = None,
+    ):
+        return super().__new__(
+            cls,
+            additions or [],
+            deletions or [],
+            modifications or [],
```

Comment:
should the `Tuple[str, str]` be some type like `DiffData` to ensure it's robust? it seems lots of the diff check logic (and therefore the reconciliation as well) will depend on these lists of tuples.

user-facing wise, `str, str` is good, but in here, we probably want to 1) `dagster.check` the types at run time, 2) better type/schematize the user inputs.

Reply:
added a type alias, and a bunch of `check` checks 👍